examples/examples/dragonquotes/dragon_queryable.rs

```diff
@@ -0,0 +1,122 @@
+use futures::prelude::*;
+use futures::select;
+use rand::{thread_rng, Rng};
+use std::convert::TryFrom;
+use std::fs::File;
+use std::io::BufRead;
+use std::io::BufReader;
+use std::num::ParseIntError;
+use std::path::Path;
+use std::time::Duration;
+use zenoh::config::Config;
+use zenoh::prelude::r#async::AsyncResolve;
+use zenoh::prelude::*;
+
+const QUOTES_INPUT_PATH: &str = "quotes.txt";
+
+const GET_ANY_PATH: &str = "zenoh/quote/any";
+const GET_QUOTE: &str = "zenoh/quote/id/*";
+
+#[async_std::main]
+async fn main() {
+    // initiate logging
+    env_logger::init();
+
+    let get_any_key_expr = KeyExpr::try_from(GET_ANY_PATH).unwrap();
+    let get_quote_key_expr = KeyExpr::try_from(GET_QUOTE).unwrap();
+
+    let quotes: Vec<String> = quotes_from_file(QUOTES_INPUT_PATH);
+
+    println!("Opening session...");
+    let session = zenoh::open(Config::default()).res().await.unwrap();
+
+    println!("Creating Queryable on '{}'...", GET_ANY_PATH);
+    let get_any_queryable = session.declare_queryable(GET_ANY_PATH).res().await.unwrap();
+
+    println!("Creating Queryable on '{}'...", GET_QUOTE);
+    let get_quote_queryable = session.declare_queryable(GET_QUOTE).res().await.unwrap();
+
+    println!("Enter 'q' to quit...");
+    let mut stdin = async_std::io::stdin();
+    let mut input = [0_u8];
+    loop {
+        select!(
+            query = get_any_queryable.recv_async() => {
+                let query = query.unwrap();
+                println!(">> [Queryable ] Received Query '{}'", query.selector());
+                let random_quote = get_random_quote(&quotes);
+                query.reply(Ok(Sample::new(get_any_key_expr.clone(), random_quote))).res().await.unwrap();
+            },
+
+            query = get_quote_queryable.recv_async() => {
+                let query = query.unwrap();
+                println!(">> [Queryable ] Received Query '{}'", query.selector());
+                if query.key_expr().is_wild() {
+                    let all_quotes = get_all_quotes(&quotes);
+                    query.reply(Ok(Sample::new(get_quote_key_expr.clone(), all_quotes))).res().await.unwrap();
+                } else {
+                    let quote = get_quote(&quotes, get_quote_number(query.key_expr()).unwrap());
+                    query.reply(Ok(Sample::new(get_quote_key_expr.clone(), quote))).res().await.unwrap();
+                }
+            },
+
+            _ = stdin.read_exact(&mut input).fuse() => {
+                match input[0] {
+                    b'q' => break,
+                    0 => sleep(Duration::from_secs(1)).await,
+                    _ => (),
+                }
+            }
+        );
+    }
+}
+
+fn get_quote_number(key_expr: &str) -> Result<usize, ParseIntError> {
+    return key_expr
+        .rfind('/')
+        .map(|i| &key_expr[i + 1..])
+        .unwrap()
+        .parse::<usize>();
+}
+
+fn get_random_quote(quotes: &Vec<String>) -> String {
+    let mut rng = thread_rng();
```

Comment:
_Extreme_ nitpick incoming, more as an excuse to point you to a few niche subjects.

`thread_rng` lazily constructs RNGs through Thread Local Storage.
In async-land, that may cause up to N constructions of RNGs, N being your number of cores, as your futures may run on any thread.

It's extremely nitpicky in this case, since the RNG is cached, but other RNG constructors may not have cached it, which would've led to this approach constructing a new RNG each time. Good constructors will use the machine's entropy generator for seeding, which could be much slower than pseudo-random generation from an already existing RNG.

Reply:
That's a nice take! How can I make the RNG to be constructed the storage shared between threads? Should I initialise it using lazy_static! putting its instance in the heap (maybe using Box?) ? Couldn't having a shared RNG cause threads to have to wait for another thread to free the resource, thus inducing some efficiency penalty?

Reply:
I mean I asked about lazy_static! because if I understand correctly it's meant to initialise lazily static variables during runtime. Do you think in this case it'd be a good use case for it?

Reply:
Having a static RNG would be worse, since you would need to synchronise them (RNG is extremely stateful by definition).

What you could do is use a closure rather than function for the callback, and enclose your RNG using `move`. That way, your closure obtains unique ownership of you RNG. It does however become a `FnMut`, which would need synchronisation to be run in parallel like Zenoh does for callbacks. But if you know your closure don't be shared, it's a good option.

In the end, `thread_local` is likely the best fit for this usecase, but you may have to do that manually for other RNGs. I just wanted to point out out because thread locals are fun